presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java

```diff
@@ -228,5 +229,44 @@ private double operate(ArithmeticBinaryExpression.Type type, double left, double
                     throw new IllegalStateException("Unsupported ArithmeticBinaryExpression.Type: " + type);
             }
         }
+
+        @Override
+        protected SymbolStatsEstimate visitCoalesceExpression(CoalesceExpression node, Void context)
+        {
+            requireNonNull(node, "node is null");
+            SymbolStatsEstimate result = null;
+            for (Expression operand : node.getOperands()) {
+                SymbolStatsEstimate operandEstimates = process(operand);
+                if (result != null) {
+                    result = estimateCoalesce(result, operandEstimates);
+                }
+                else {
+                    result = operandEstimates;
+                }
+            }
+            return requireNonNull(result, "result is null");
+        }
+
+        private SymbolStatsEstimate estimateCoalesce(SymbolStatsEstimate left, SymbolStatsEstimate right)
+        {
+            // Question to reviewer: do you have a method to check if fraction is empty or saturated?
+            if (left.getNullsFraction() == 0) {
+                return left;
+            }
+            else if (left.getNullsFraction() == 1.0) {
+                return right;
+            }
+            else {
+                return SymbolStatsEstimate.builder()
+                        .setLowValue(min(left.getLowValue(), right.getLowValue()))
+                        .setHighValue(max(left.getHighValue(), right.getLowValue()))
```

Comment:
`right.getHighValue()`?

Reply:
Good catch! I added a test for that too.